File: absensi-mahasiswa/src/utils/searching.js
// Sequential Search untuk mencari mahasiswa berdasarkan nama (tanpa break/continue)
export const sequentialSearchByName = (data, searchName) => {
  const results = [];
  const searchLower = searchName.toLowerCase();
  
  if (searchLower.length === 0) {
    return data;
  }
  
  for (let i = 0; i < data.length; i++) {
    const nameLower = data[i].name.toLowerCase();
    let isMatch = true;
    let j = 0;
    
    while (j < searchLower.length && isMatch) {
      if (j >= nameLower.length || nameLower[j] !== searchLower[j]) {
        isMatch = false;
      }
      j = j + 1;
    }
    
    if (isMatch) {
      results.push(data[i]);
    }
  }
  
  return results;
};

// Binary Search untuk mencari absensi berdasarkan tanggal (data harus terurut)
export const binarySearchByDate = (data, searchDate) => {
  let left = 0;
  let right = data.length - 1;
  let result = null;
  let found = false;
  
  while (left <= right && !found) {
    const mid = Math.floor((left + right) / 2);
    
    if (data[mid].date === searchDate) {
      result = data[mid];
      found = true;
    } else if (data[mid].date < searchDate) {
      left = mid + 1;
    } else {
      right = mid - 1;
    }
  }
  
  return result;
};

// Sort data sebelum binary search
export const sortAttendanceByDate = (data) => {
  const arr = [...data];
  const n = arr.length;
  
  for (let i = 0; i < n - 1; i++) {
    for (let j = 0; j < n - i - 1; j++) {
      if (arr[j].date > arr[j + 1].date) {
        const temp = arr[j];
        arr[j] = arr[j + 1];
        arr[j + 1] = temp;
      }
    }
  }
  
  return arr;
};

File: backend-absensi/utils/searching.go
package utils

import (
	"backend-absensi/models"
)

func SequentialSearchByName(students [50]models.Student, count int, searchName string) []models.Student {
	var results []models.Student

	if searchName == "" {
		return results
	}

	for i := 0; i < count; i++ {
		isMatch := true
		searchLower := toLower(searchName)
		nameLower := toLower(students[i].Name)

		if len(searchLower) > len(nameLower) {
			isMatch = false
		} else {
			for j := 0; j < len(searchLower) && isMatch; j++ {
				if nameLower[j] != searchLower[j] {
					isMatch = false
				}
			}
		}

		if isMatch {
			results = append(results, students[i])
		}
	}

	return results
}

func BinarySearchByDate(attendances []models.Attendance, searchDate string) *models.Attendance {
	sorted := make([]models.Attendance, len(attendances))
	copy(sorted, attendances)

	insertionSortAttendanceByDate(sorted)

	left := 0
	right := len(sorted) - 1
	var result *models.Attendance = nil
	found := false

	for left <= right && !found {
		mid := (left + right) / 2

		if sorted[mid].Date == searchDate {
			result = &sorted[mid]
			found = true
		} else if sorted[mid].Date < searchDate {
			left = mid + 1
		} else {
			right = mid - 1
		}
	}

	return result
}

func toLower(s string) string {
	result := ""
	for _, c := range s {
		if c >= 'A' && c <= 'Z' {
			result += string(c + 32)
		} else {
			result += string(c)
		}
	}
	return result
}
